lib/bridge/PowerlevelSyncer.js

```diff
@@ -0,0 +1,402 @@
+        if (modified === 0) {
+            // We didn't actually change anything, so don't send anything.
+            return;
+        }
+        const botClient = this._ircBridge.getAppServiceBridge().getIntent().getClient();
+        await botClient.sendStateEvent(roomId, "m.room.power_levels", plContent, "");
+    }
+
+    /**
+     * If a mode given in PRIVATE_MODES is found, change a room's join rules
+     * to match.
+     * @param {BridgeReqeust} req The request tracking the operation.
+     * @param {IrcServer} server The server the channel and users are part of
+     * @param {string} channel Which channel was the mode(s) set in.
+     * @param {string} mode The mode string.
+     * @param {boolean} enabled Was the mode enabled or disabled.
+     */
+    async _onPrivateMode(req, server, channel, mode, enabled) {
+        // 'k' = Channel requires 'keyword' to join.
+        // 'i' = Channel is invite-only.
+        // 's' = Channel is secret
+
+        // For k and i, we currently want to flip the join_rules to be
+        // 'invite' to prevent new people who are not in the room from
+        // joining.
+
+        // For s, we just want to control the room directory visibility
+        // accordingly. (+s = 'private', -s = 'public')
+
+        // TODO: Add support for specifying the correct 'keyword' and
+        // support for sending INVITEs for virtual IRC users.
+        let matrixRooms = await this._ircBridge.getStore().getMatrixRoomsForChannel(
+            server, channel
+        );
+        if (matrixRooms.length === 0) {
+            req.log.info("No mapped matrix rooms for IRC channel %s", channel);
+            return;
+        }
+
+        if (mode === "s") {
+            if (!server.shouldPublishRooms()) {
+                req.log.info("Not syncing publicity: shouldPublishRooms is false");
+                return;
+            }
+            const key = this._ircBridge.publicitySyncer.getIRCVisMapKey(
+                server.getNetworkId(), channel
+            );
+
+            matrixRooms.map((room) => {
+                this._ircBridge.getStore().setModeForRoom(room.getId(), "s", enabled);
+            });
+            // Update the visibility for all rooms connected to this channel
+            await this._ircBridge.publicitySyncer.updateVisibilityMap(
+                true, key, enabled
+            );
+        }
+        // "k" and "i"
+        matrixRooms.map((room) => {
+            this._ircBridge.getStore().setModeForRoom(room.getId(), mode, enabled);
+        });
+
+        const promises = matrixRooms.map((room) => {
+            switch (mode) {
+                case "k":
+                case "i":
+                    req.log.info((enabled ? "Locking room %s" :
+                        "Reverting %s back to default join_rule"),
+                        room.getId()
+                    );
+                    if (enabled) {
+                        return this._setMatrixRoomAsInviteOnly(room, true);
```

Comment:
it's a bit unfortunate that the powerlevel syncer also affects access to the room. Can we just call this a `AccessSyncer` instead or something?